src/v0/destinations/tiktok_ads_offline_events/transform.js

```diff
@@ -0,0 +1,261 @@
+const { SHA256 } = require("crypto-js");
+const set = require("set-value");
+const _ = require("lodash");
+const { EventType } = require("../../../constants");
+const {
+  constructPayload,
+  defaultRequestConfig,
+  removeUndefinedAndNullValues,
+  isDefinedAndNotNullAndNotEmpty,
+  getHashFromArrayWithDuplicate,
+  checkInvalidRtTfEvents,
+  handleRtTfSingleEventError,
+  getSuccessRespEvents,
+  defaultBatchRequestConfig
+} = require("../../util");
+const {
+  CONFIG_CATEGORIES,
+  MAX_BATCH_SIZE,
+  MAPPING_CONFIG,
+  EVENT_NAME_MAPPING,
+  PARTNER_NAME
+} = require("./config");
+const {
+  ConfigurationError,
+  InstrumentationError
+} = require("../../util/errorTypes");
+
+const getContents = message => {
+  const contents = [];
+  const { properties } = message;
+  const { products } = properties;
+  const mappingJson =
+    MAPPING_CONFIG[CONFIG_CATEGORIES.TRACK_PROPERTIES_CONTENTS.name];
+
+  const updateContentsArray = productProps => {
+    const contentObject = constructPayload(productProps, mappingJson);
+    const contentObj = removeUndefinedAndNullValues(contentObject);
+    if (Object.keys(contentObj).length > 0) {
+      contents.push(contentObj);
+    }
+  };
+
+  if (products && Array.isArray(products) && products.length > 0) {
+    products.forEach(product => updateContentsArray(product));
+  } else {
+    updateContentsArray(properties);
+  }
+  return contents;
+};
+
+const getTrackResponse = (message, category, Config, event) => {
+  const { hashUserProperties, accessToken } = Config;
+  const payload = constructPayload(message, MAPPING_CONFIG[category.name]);
+  payload.event = event;
+  payload.partner_name = PARTNER_NAME;
+
+  // Settings Contents object
+  const contents = getContents(message);
+  if (contents.length > 0) {
+    set(payload, "properties.contents", contents);
+  }
+
+  const email = message?.traits?.email || message?.context?.traits.email;
+  if (isDefinedAndNotNullAndNotEmpty(email)) {
+    const emails = hashUserProperties
```

Comment:
can't we do this through `mapping.json` ? Any restrictions we saw while trying to use `mapping.json` ?!

Reply:
we need to hash the email(depending on config setting) and end it in an array format. So added the field here rather than adding it in mapping config and again transforming it.

Reply:
You can still convert using mapping.json & then make it into an array would've been easier as internal implementations would be hashed as per the algo we had in-place already(basically we would be utilising the common functions more). It is fine for now